### PR TITLE
refactor: avoid unnecessary rerender of tabled components & typing

### PR DESCRIPTION
**Pull request by ignromanov**
## Summary

- typing table data usage
- memoize all tables data
- memoize tables callbacks

## Checklist

-   [x] My changes look good in both light AND dark mode
-   [x] The change is not hardcoded to a single network, but has multi-asset in mind
-   [x] I checked my changes for obvious issues, debug statements and commented code
-   [x] Tests _(if necessary)_
-   [x] Ready to be merged

### CODE REVIEW

File: src/domains/dashboard/components/Wallets/Wallets.contracts.tsx [hunk 1]
```diff
@@ -1,9 +1,20 @@
 import { Contracts } from "@payvo/profiles";
-import { DropdownOption } from "app/components/Dropdown";
+
+import { DropdownOption } from "../../../../app/components/Dropdown";
```
Comment (w3ea) [inline src/domains/dashboard/components/Wallets/Wallets.contracts.tsx:3]:
```suggestion
import { DropdownOption } from "app/components/Dropdown";
```

File: src/domains/plugin/components/PluginUpdatesConfirmation/PluginUpdatesConfirmation.contracts.tsx [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { ExtendedSerializedPluginConfigurationData } from "../../../../plugins";
```
Comment (w3ea) [inline src/domains/plugin/components/PluginUpdatesConfirmation/PluginUpdatesConfirmation.contracts.tsx:1]:
```suggestion
import { ExtendedSerializedPluginConfigurationData } from "plugins";
```

File: src/domains/transaction/components/UnlockTokens/blocks/UnlockTokensSelect/UnlockTokensSelect.tsx [hunk 3]
```diff
@@ -1,39 +1,31 @@
...
-import React, { useEffect, useMemo, useState } from "react";
+import React, { FC, useCallback, useEffect, useMemo, useState } from "react";
 import { useFormContext } from "react-hook-form";
 import { useTranslation } from "react-i18next";
+import { Column } from "react-table";
 
-import { UnlockableBalance, UnlockableBalanceSkeleton, UnlockTokensFormState } from "../../UnlockTokens.contracts";
-import { useColumns } from "../../UnlockTokens.helpers";
+import { Checkbox } from "../../../../../../app/components/Checkbox";
+import { Tooltip } from "../../../../../../app/components/Tooltip";
```
Comment (w3ea) [inline src/domains/transaction/components/UnlockTokens/blocks/UnlockTokensSelect/UnlockTokensSelect.tsx:12]:
```suggestion
import { Checkbox } from "app/components/Checkbox";
import { Tooltip } from "app/components/Tooltip";
```

File: src/domains/wallet/pages/ImportWallet/Ledger/LedgerTabs.contracts.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { Contracts } from "@payvo/profiles";
+import { Networks } from "@payvo/sdk";
+
+import { useLedgerScanner } from "../../../../../app/contexts";
```
Comment (w3ea) [inline src/domains/wallet/pages/ImportWallet/Ledger/LedgerTabs.contracts.tsx:4]:
```suggestion
import { useLedgerScanner } from "app/contexts";
```

File: src/tests/mocks/StubStorage/StubStorage.test.ts [hunk 5]
```diff
@@ -20,7 +20,7 @@ describe("StubStorage", () => {
 	it("should return set and get an entry into storage", async () => {
 		await stubStorage.set("item", "bleh");
 
-		await expect(stubStorage.get("item")).tresolves.oEqual("bleh");
+		await expect(stubStorage.get("item")).resolves.toEqual("bleh");
```
Comment (w3ea) [inline src/tests/mocks/StubStorage/StubStorage.test.ts:23]: [already fixed here](https://github.com/PayvoHQ/wallet/pull/539/files#diff-7a8327de1ea68669135d7f7b6d05c7a657e414701081258e73f778b6874ae706R23)

File: src/domains/contact/pages/Contacts/Contacts.tsx [hunk 6]
```diff
@@ -5,24 +5,21 @@ import { Header } from "app/components/Header";
...
-interface ContactsHeaderExtraProperties {
-	showSearchBar: boolean;
-	onSearch?: any;
-	onAddContact?: any;
-}
+import { ContactListItemOption } from "../../components/ContactListItem/ContactListItem.models";
+import { ContactsHeaderExtraProperties } from "./Contacts.contracts";
 
-const ContactsHeaderExtra = ({ showSearchBar, onSearch, onAddContact }: ContactsHeaderExtraProperties) => {
+const ContactsHeaderExtra: FC<ContactsHeaderExtraProperties> = ({ showSearchBar, onSearch, onAddContact }) => {
```
Comment (w3ea) [inline src/domains/contact/pages/Contacts/Contacts.tsx:22]: `React.FC` always imply children, even if we don't want it (https://github.com/facebook/create-react-app/pull/8177)
it's better to remove it (also [here](https://github.com/PayvoHQ/wallet/pull/540/files#diff-a1c8c12dedb1f2ebfcd836a01b5fff46a23237eee1f20a40b0ebf3c9aa891e3aR15), [here](https://github.com/PayvoHQ/wallet/pull/540/files#diff-0f1ff3873311e16639d84dab8635ff7372061ffa6182e5f036ba74a12a49998dR11), ...)

[this is a good article](https://fettblog.eu/typescript-react-why-i-dont-use-react-fc/) about the downsides of `React.FC`

File: src/domains/transaction/components/RecipientList/RecipientList.tsx [hunk 7]
```diff
@@ -113,54 +114,53 @@ const RecipientListItem = ({
...
-const defaultProps = {
-	recipients: [],
-};
-
-export const RecipientList = ({
+export const RecipientList: FC<RecipientListProperties> = ({
 	assetSymbol,
 	isEditable = false,
-	recipients = defaultProps.recipients,
+	recipients = [],
```
Comment (w3ea) [inline src/domains/transaction/components/RecipientList/RecipientList.tsx:120]: placing the array, object, or function inside a component as DefaultProps [causes unwanted re-renders](https://github.com/PayvoHQ/wallet/issues/340#:~:text=Keep%20in%20mind%20that%20the%20default%20object%20should%20be%20stored%20in%20a%20constant%20outside%20of%20the%20component/function%20or%20you%20can%20cause%20unwanted%20re-renders%20because%20the%20object%20gets%20recreated%20on%20every%20call/mount%20if%20it%20isn%27t%20stored%20outside%20of%20them.) (especially in the complex components)

File: src/domains/transaction/components/RecipientList/RecipientList.tsx [hunk 8]
```diff
@@ -113,54 +114,53 @@ const RecipientListItem = ({
...
+			{ Header: "Address", className: "hidden" },
+			...(showAmount ? [{ Header: "Amount", className: "hidden" }] : []),
+			...(isEditable ? [{ Header: "Action", className: "hidden" }] : []),
+		],
+		[showAmount, isEditable],
+	);
 
-	if (showAmount) {
-		columns.push({ Header: "Amount", className: "hidden" });
-	}
```
Comment (w3ea) [inline src/domains/transaction/components/RecipientList/RecipientList.tsx:138]: i think the previous codes are simpler and have slightly better performance (the new codes execute else condition)
[also here](https://github.com/PayvoHQ/wallet/pull/540/files#diff-6f13e409986a8d358ce1bdf3de39dff61ec9c4f8f3ecc31c54f403e3bc7a06e8R32-R54)

File: src/domains/transaction/components/TransactionTable/PendingTransactionsTable/PendingTransactionsTable.tsx [hunk 9]
```diff
@@ -20,7 +20,67 @@ export const PendingTransactions = ({
...
+			{
+				Header: t("COMMON.ID"),
+				minimumWidth: true,
+			},
+			{
+				Header: t("COMMON.DATE"),
+				accessor: () => "timestamp",
+				cellWidth: "w-50",
+				sortDescFirst: true,
+			},
```
Comment (w3ea) [inline src/domains/transaction/components/TransactionTable/PendingTransactionsTable/PendingTransactionsTable.tsx:34]: it's better to keep the columns out of components to make them cleaner and more compact
there is an [ESLint rule](https://github.com/PayvoHQ/wallet/blob/develop/.eslintrc.js#L116) for this

File: src/domains/contact/pages/Contacts/Contacts.tsx (hunk 6 above)
Comment (ignromanov) [reply src/domains/contact/pages/Contacts/Contacts.tsx:22]: That's an interesting point about using FC. However, what I like to use it for:
1. When defining a type in advance, the IDE immediately detects and shows the types of the parameters being passed, this does not happen with post-definition type.
2. It keeps track of the correct type of returned values.
3. To not take children by default, we can use `VFC`.

image

Comment (w3ea) [reply src/domains/contact/pages/Contacts/Contacts.tsx:22]: for the no. 1, there is no different for me (i'm using WebStorm) :hmhm:

imo [VoidFunctionComponent (VFC)](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f4f453b018d0d512cde50efb6db35fd3f8a001ab/types/react/index.d.ts#L551) is a weird name for components that return JSX
but in general, we can use VFC here

File: src/domains/transaction/components/RecipientList/RecipientList.tsx (hunk 7 above)
Comment (ignromanov) [reply src/domains/transaction/components/RecipientList/RecipientList.tsx]: restored

File: src/domains/transaction/components/RecipientList/RecipientList.tsx (hunk 8 above)
Comment (ignromanov) [reply src/domains/transaction/components/RecipientList/RecipientList.tsx]: Ok, I restored the If variants

File: src/domains/transaction/components/TransactionTable/PendingTransactionsTable/PendingTransactionsTable.tsx (hunk 9 above)
Comment (ignromanov) [reply src/domains/transaction/components/TransactionTable/PendingTransactionsTable/PendingTransactionsTable.tsx]: Restored and move some columns declarations outside of the components

File: src/domains/contact/pages/Contacts/Contacts.tsx (hunk 6 above)
Comment (ignromanov) [reply src/domains/contact/pages/Contacts/Contacts.tsx]: `VFC` is better than specifying types as in a usual function. I will use it instead of `FC`

**Review by faustbrian — APPROVED** [review]

**Pull request merged**